### ISSUE TITLE
Segmentation Fault when using a custom item

### ISSUE THREAD

**Issue by kelvinhammond** [openzim/node-libzim#80]
In libzim v8 when using a custom item and adding a lot of objects to a zim file it will throw a segmentation fault error after about 10,000 items have been inserted.

The error has been reported to  [nodejs/node-addon-api](https://github.com/nodejs/node-addon-api) below. 

It appears to be related to:
- the persistent object reference being lost or garbage collected before the object and function can be used
- some other unknown issue involving the global state within nodejs.
- it may also be related to the creation of ThreadSafeFunction(s) created within a ThreadSafeFunction context but this should be OK and does not crash until after about 10k or more iterations.

To reproduce the error modify `tests/makeLargeZim.ts` file to use the custom item and run the test, it should fail.

The issue has been reported to https://github.com/nodejs/node-addon-api/issues/1174 .

Related to https://github.com/openzim/node-libzim/pull/72

**Comment by kelson42** [comment]
@kelvinhammond One hour ago, I have updated the build script to use libzim `8.1.0` which has been released yesterday (and fix/update the CI). I have also run (for the first time) `$ npm run test-mem-leak` twice on my Ubuntu 22.04 + Node.js 18.12.1 and... all works fine. It ran twice up to the end without any error message. Maybe it would be worth it you try again?

**Comment by kelson42** [comment]
@kelvinhammond Thx, would be really helpful to know soon if this critical bug is really fixed or if I just don't run the reproduction case steps properly.

**Comment by kelvinhammond** [comment]
> @kelvinhammond Thx, would be really helpful to know soon if this critical bug is really fixed or if I just don't run the reproduction case steps properly.

@kelson42 Try it with these changes
Edit `test/makeLargeZim.ts`
Uncomment the `const customItem` code and have `await creator.addItem(stringItem);` be replaced by `await creator.addItem(customItem);` and see if that works for you.

The diff looks like this.
```
diff --git a/test/makeLargeZim.ts b/test/makeLargeZim.ts
index b5db866..e87979e 100644
--- a/test/makeLargeZim.ts
+++ b/test/makeLargeZim.ts
@@ -43,13 +43,13 @@ console.log(`Making ZIM file with [${numArticles}] articles`);
       data,
     );

-    /*
     const customItem = { // custom item
       path: url,
       mimeType: "text/html",
       title: title,
       hints: {FRONT_ARTICLE: 1},
       getContentProvider() { // custom content provider
+        return stringItem.getContentProvider();
         let sent = false;
         return {
           size: data.length,
@@ -63,9 +63,9 @@ console.log(`Making ZIM file with [${numArticles}] articles`);
         };
       },
     };
-    */

-    await creator.addItem(stringItem);
+    //await creator.addItem(stringItem);
+    await creator.addItem(customItem);
   }

   console.log(`Finalising...`);
```

**Comment by kelvinhammond** [comment]
I tested it with node v19.1.0

I get 
```
terminate called after throwing an instance of 'Napi::Error'

  what():  Illegal invocation

#8  0x00007ffff031a5c3 in Napi::Function::Call (this=0x7fffffffa0c0, recv=0x555555632990, argc=0, args=0x0) at 
/node-libzim/node_modules/node-addon-api/napi-inl.h:2341

#9  0x00007ffff031a482 in Napi::Function::Call (this=0x7fffffffa0c0, recv=0x555555632990, args=std::initializer_list of length 0) at /node-libzim/node_modules/node-addon-api/napi-inl.h:2306

#10 0x00007ffff0326f92 in ContentProviderWrapper::feed()::{lambda(Napi::Env, Napi::Function)#1}::operator()(Napi::Env, Napi::Function) const (__closure=0x7fffc40245f0, env=..., feedFunc=...) at ../src/contentProvider.h:79
```

Looking into this more and trying it with a newer version of node-addon-api

**Comment by kelvinhammond** [comment]
@kelson42 It does work, just not with custom items.

**Comment by kelson42** [comment]
@kelvinhammond Sorry for the big delay in answering. I have involved @mgautierfr who is the lead developer of the libzim. Could you please elaborate what are these "custom items". This is not a concept of the libzim and therefore this is a bit puzzling. One important other point is if we really need them (does that impact MWoffliner)?

**Comment by kelson42** [comment]
BTW @uriesk, this is the bug which stops us to use latest libzim in MWoffliner. If you like to be challenged, this might be one for you ;)

**Comment by kelvinhammond** [comment]
@kelson42 In libzim you can create Item(s) which have each have a `ContentProvider`, some are provided by default such as `StringItem` and `FileItem` but you can define and implement your own. In node-libzim `StringItem` and `FileItem` work fine but when you try to define your own it fails because of an [issue (node-addon-api#1174)](https://github.com/nodejs/node-addon-api/issues/1174) in node or node-addon-api (napi) when creating a large file (sometimes a smaller one too)

**Comment by kelson42** [comment]
@kelvinhammond But we don't need this customItem in MWoffliner? You confirm?

**Comment by kelvinhammond** [comment]
> @kelvinhammond But we don't need this customItem in MWoffliner? You confirm?

I'll test a large file using only StringItem and FileItem to confirm it works but yes, mwoffliner does not use the custom item and content providers

**Comment by kelson42** [comment]
@kelvinhammond This would be a good news. Waiting to your confirmation. If confirmed, we could consider to release with this bug/limitation.

**Comment by uriesk** [comment]
> BTW @uriesk, this is the bug which stops us to use latest libzim in MWoffliner (we are stuck with libzim6 which is18 months old). If you like to be challenged, this might be one for you ;)

sorry, it's out of my league.
I am not yet familiar enough with node-addon-api.

**Comment by kelson42** [comment]
> @kelson42 In libzim you can create Item(s) which have each have a `ContentProvider`, some are provided by default such as `StringItem` and `FileItem` but you can define and implement your own. In node-libzim `StringItem` and `FileItem` work fine but when you try to define your own it fails because of an [issue (node-addon-api#1174)](https://github.com/nodejs/node-addon-api/issues/1174) in node or node-addon-api (napi) when creating a large file (sometimes a smaller one too)

@mgautierfr Do you confirm this (not sure this is exactly how things work at the libzim)? Maybe there is a few things to know and check if using a custom provider?

**Comment by mgautierfr** [comment]
One important thing to remember is that the ContentProvider returned by `getContentProvider` must stay valid after libzim has release the `shared_ptr<Item>` passed in `creator::addItem`.

libzim will use the `shared_ptr<Item>` to get a `ContentProvider` but will release the shared ptr before using the `ContentProvider` (potentially in another thread). So if  `ContentProvider` is referencing the data stored in the item. We may have a situation where we free the item (and so, its data) and we have a `ContentProvider` referencing a freed data.
I don't know how `item_.Value()` behaves. But if it return a reference/pointer to the value we are in this situation.
`ContentProvider` must prevent the value it will return from being freed. Either by owning the data, copying it, keeping a reference to the item containing the data, ...

I don't know how napi wrap things. But here : https://github.com/openzim/node-libzim/blob/main/test/makeLargeZim.ts#L54-L62, your custom item return a contentProvider which contain a closure as `feed` method.
So you must be sure that the referenced data (literally `data`) is not freed when libzim is calling `feed` method.
As libzim keep a reference to the `contentProvider` without keeping a reference `data` **maybe** NAPI think that nothing reference the `data` variable and free it.
And depending of how closures work in js, it can be worse as all `contentprovider` may share the same data.
At least in python, this is the case (see my own old answer in stackoverflow at this subject : https://stackoverflow.com/questions/21449176/issue-with-generating-ttk-checkboxes-in-loops-and-passing-arguments/21459922#21459922)

Maybe you can try this custom item which not create closure (but copy (reference to?) data):

```javascript
const customItem = { // custom item
  path: url,
  mimeType: "text/html",
  title: title,
  hints: {FRONT_ARTICLE: 1},
  item_data: data, // reference directly the data without closure
  getContentProvider() { // custom content provider
    return {
      size: this.item_data.length,
      data_to_send: this.item_data, // same here, we don't want the feed() method being a closure accessing the `item_data` in customItem which will be deleted.
      sent = false,
      feed() {
        if(!this.sent) {
          this.sent = true;
          return new Blob(this.data_to_send);
        }
        return new Blob();
      }
    };
  },
};
```

**Comment by kelson42** [comment]
@kelvinhammond Does that help? Do we have anything preventing the data to be freed?

**Comment by kelvinhammond** [comment]
@kelson42 Sadly, it did not. I store a reference to the objects and functions in node so it "should" keep a copy of it in memory without garbage collecting it. The error also seems to happen (most of the time) when the object is creating a thread safe function callback.

As for the `StringItem`, that works, I haven't verified `FileItem` yet (I need to write a bit of code to test this too in a large tmp folder for each article we store). `StringItem` and `FileItem` work because they don't encounter the bug calling to `getContentProvider` and `feed` because I use an internally stored `StringItem` and `FileItem` smart ptr. I can test `FileItem` to be sure but I think it's ready to go. Is there a way to test mwoffliner with a large enough (doesn't need to be huge) archive to verify it works?

**Comment by kelson42** [comment]
> Is there a way to test mwoffliner with a large enough (doesn't need to be huge) archive to verify it works?

You mean with a big wikipedia?

**Comment by kelson42** [comment]
Considering that this bug does not impact our current scrapers AFAIK, I propose to release 3.0.0 with this bug, so we can move forward and benefit from all the work already done.

**Comment by kelvinhammond** [comment]
> Considering that this bug does not impact our current scrapers AFAIK, I propose to release 3.0.0 with this bug, so we can move forward and benefit from all the work already done.
> 
> @kelvinhammond @mgautierfr Any thpughts?

Sure, release it but include a note in the change notes please.

**Comment by benoit74** [comment]
@kelvinhammond are you by any chance adding items to the creator in parallel ? See #155

**Comment by kelvinhammond** [comment]
It's been awhile, I'll see if this is still an issue.

**Comment by kelvinhammond** [comment]
So I did take a look at this and I still can't quite figure out where the issue is.

I noticed a few things we could do such as:
1. Adding a mutex between calls to `creator_->addItem` which may not be thread safe.
2. Same above but for `finishZimCreation`

I tried tracking the number of threads and using that during initialization of the `Napi::ThreadSafeFunction`s but that didn't seem to do much.

I also tried using the `ThreadSafeFunction::NonBlockingCall` to see if we didn't have enough calls and still nothing.

I also upgraded to `node-addon-api` version `8.5.0` but no improvement.

As I currently understand it, based on some testing and looking at the libzim codebase.
libzim's `Creator` defaults to 4 threads, I changed it to 1 for testing and no fix because it's basically 2 threads minimum.

I have to use threads because of the following.  
`creator.addItem` called from the main javascript thread will eventually wait for a cluster to finish writing to disc and close before starting the next one, when it does that that cluster will call; from another thread, `getContentProvider` and / or `contentProvider.feed` which will call a `ThreadSafeFunction` back in javascript land but it wait's on the main thread which we're currently in thus causing a deadlock.

To work around this I put the `creator.addItem` call into an `AsyncWorker` (a separate thread with a promise tied to it) so it won't block the main thread, you'd have to wait on the promise to resolve in javascript land or else you might overload the memory with AsyncWorker threads though.

The problem I face is inconsistent errors for one.
It will sometimes crash with no error, other times I get an error of

```
# Fatal error in , line 0
# Check failed: node->IsInUse().
#
#
#
#FailureMessage Object: 0x7fffa85a1d60
----- Native stack trace -----

 1: 0x632e5f0b32a7  [/usr/bin/node]
 2: 0x632e6034918f V8_Fatal(char const*, ...) [/usr/bin/node]
 3: 0x632e5f48f83e v8::internal::GlobalHandles::NodeSpace<v8::internal::GlobalHandles::Node>::Release(v8::internal::GlobalHandles::Node*) [/usr/bin/node]
 4: 0x632e5ef00129 node::AsyncResource::~AsyncResource() [/usr/bin/node]
 5: 0x632e5efccb5f napi_delete_async_work [/usr/bin/node]
 6: 0x76b9c0317fd4 Napi::AsyncWorker::~AsyncWorker() [/node-libzim/build/Debug/zim_binding.node]
 7: 0x76b9c032ab8c AddItemAsyncWorker::~AddItemAsyncWorker() [/node-libzim/build/Debug/zim_binding.node]
 8: 0x76b9c032aba8 AddItemAsyncWorker::~AddItemAsyncWorker() [/node-libzim/build/Debug/zim_binding.node]
 9: 0x76b9c0318064 Napi::AsyncWorker::Destroy() [/node-libzim/build/Debug/zim_binding.node]
10: 0x76b9c031878c Napi::AsyncWorker::OnWorkComplete(Napi::Env, napi_status) [/node-libzim/build/Debug/zim_binding.node]
11: 0x76b9c03185cc Napi::AsyncWorker::OnAsyncWorkComplete(napi_env__*, napi_status, void*) [/node-libzim/build/Debug/zim_binding.node]
12: 0x632e5efce000  [/usr/bin/node]
13: 0x632e5efd0787 node::ThreadPoolWork::ScheduleWork()::{lambda(uv_work_s*, int)#2}::operator()(uv_work_s*, int) const [/usr/bin/node]
14: 0x76b9c7163efb  [/usr/lib/libuv.so.1]
15: 0x76b9c7159bc3  [/usr/lib/libuv.so.1]
16: 0x76b9c7179af3  [/usr/lib/libuv.so.1]
17: 0x76b9c7160252 uv_run [/usr/lib/libuv.so.1]
18: 0x632e5ef02f57 node::SpinEventLoopInternal(node::Environment*) [/usr/bin/node]
19: 0x632e5f0752db node::NodeMainInstance::Run(node::ExitCode*, node::Environment*) [/usr/bin/node]
20: 0x632e5f07586e node::NodeMainInstance::Run() [/usr/bin/node]
21: 0x632e5efc65d4 node::Start(int, char**) [/usr/bin/node]
22: 0x76b9c5e2de08  [/usr/lib/libc.so.6]
23: 0x76b9c5e2decc __libc_start_main [/usr/lib/libc.so.6]
24: 0x632e5eeff615 _start [/usr/bin/node]
```

This is the command I test with, I did uncomment the custom `contentProvider` code in the file `test/makeLargeZime.ts`.
> `node-gyp rebuild --debug -v && npm run bundle && npm run test-mem-leak --force-node-api-uncaught-exceptions-policy=true`

**Comment by kelvinhammond** [comment]
Any help with the above would be much appreciated.

**Comment by benoit74** [comment]
I've just opened https://github.com/openzim/node-libzim/issues/164 yesterday which confirms there is an issue wider than what @kelvinhammond is experiencing, and indeed this becomes my P0 now, it is just too annoying in mwoffliner to have these random segfault.

I suspect we have two distinct issues.

First, @kelvinhammond you are right that you need a mutex to protect calls to `creator->addItem`. This is not thread-safe, but expect to return very quickly, so adding a mutex is not expected to hurt at all. Basically it just transfer the info of what needs to be added to the libzim, and then libzim takes care of doing required work with the 4 threads you've already mentioned (good job you've find them).

I do not expect that adding a mutex to protect call to `finishZimCreation` will do anything because this is supposed to be called only once, when you've already finished to add all items. Of course, I might be missing something. This is a blocking method which will wait for the 4 threads to finish their work and then do required final work. You should not continue to add items once you've called `finishZimCreation`, but I suspect this is kinda obvious to you.

Second, and this is linked to the other issue #164, I'm now quite convinced we have an issue in our content providers. In many cases we pass only a pointer to the content to the libzim and, as mgautier, I suspect that Napi / Node.JS fails to follow correctly when this pointer can be released, and release underlying memory before `creator->addItem` has finished its job. In most cases this is invisible because libzim achieves to complete addition before Node.JS VM triggers the memory release. But sometimes it is not. This is probably significantly depending on the size of the item we are adding, so I will work in this direction to have a reliably failing test.

**Comment by benoit74** [comment]
@kelvinhammond I just had a look at the log of https://github.com/openzim/node-libzim/issues/80#issuecomment-3085690112 and this does not look like a segfault anymore, or do I miss something? This issue was about segfault, so I'm a bit confused. Do you have a second problem? If so, please open a distinct issue (even if we finally end-up with a conclusion that problem was identical) so that we properly isolate symptoms and discussions, it is already a complex issue, so better not to track two different symptoms at once ^^

**Comment by kelvinhammond** [comment]
@benoit74 with version 8.5.0 of node-addon-api and the extra command line argument I was able to get a better error message. It's still the same issue but node catches it now.
